rosbridge_server/scripts/rosbridge_websocket.py

```diff
@@ -332,31 +326,48 @@ def protocol_parameter_handling(self):
         CallService.services_glob = RosbridgeWebSocket.services_glob
 
 
-def main(args=None):
-    if args is None:
-        args = sys.argv
+def start_ros_thread(node, shutdown_event):
+    def spin_ros():
+        executor = rclpy.executors.SingleThreadedExecutor()
+        executor.add_node(node)
+        while rclpy.ok() and not shutdown_event.is_set():
+            executor.spin_once(timeout_sec=1.0)
+        node.destroy_node()
+        rclpy.shutdown()
+
+    ros_thread = threading.Thread(target=spin_ros, daemon=True)
+    ros_thread.start()
+    return ros_thread
 
-    rclpy.init(args=args)
+
+async def async_main():
+    rclpy.init(args=sys.argv, signal_handler_options=rclpy.signals.SignalHandlerOptions.NO)
     node = RosbridgeWebsocketNode()
+    shutdown_event = threading.Event()
+    ros_thread = start_ros_thread(node, shutdown_event)
 
-    executor = rclpy.executors.SingleThreadedExecutor()
-    executor.add_node(node)
+    loop = asyncio.get_running_loop()
+    stop_event = asyncio.Event()
+    signal_handled = False
 
-    def spin_ros():
-        executor.spin_once(timeout_sec=0.01)
-        if not rclpy.ok():
-            shutdown_hook()
-
-    spin_callback = PeriodicCallback(spin_ros, 1)
-    spin_callback.start()
-    try:
-        start_hook()
-        node.destroy_node()
-        rclpy.shutdown()
-    except KeyboardInterrupt:
+    def handle_signal():
+        nonlocal signal_handled
+        if signal_handled:
+            return
         print("Exiting due to SIGINT")
-    finally:
-        shutdown_hook()  # shutdown hook to stop the server
+        shutdown_event.set()
+        stop_event.set()
+        signal_handled = True
+
+    for sig in (signal.SIGINT, signal.SIGTERM):
+        loop.add_signal_handler(sig, handle_signal)
+
+    await stop_event.wait()
+    ros_thread.join(timeout=2.0)
```

Comment:
Same here re: hard-coded timeout

Reply:
This timeout just needs to be larger than the spin_once timeout